Add reverse scan  support

thundercw opened the pull request:
The work is still in process.The main logic has been added.There are some tests and comments to be added.
The major changes are as follows:
1.The `key.Prev()` has been restored which has been deleted by @petermattis in 5ebdffa . But it still have a problem. In some cases `key.Next().Prev() != key`, for example key is end by `0xff`. So the `key.Prev()` I only used in internal look up ranges to find previous ranges.
2. Support prefetch ranges reverse by adding a reverse field to the existing `InternalRangeLookupRequest`. (#1585)
3. Add two fields `inclusiveKey` and `inclusiveEndKey` in the `ReverseScanRequest` to support cross range scan.

Please provide your suggestions. Thanks.

thundercw committed:
support reverse scan

bug fix in getCachedRangeDescriptor

add some test case

bug fix: lookup range descriptor use the key in the minddle of the range.

Errors fix after rebase

update per comments

tamird commented on storage/engine/mvcc.go:
```diff
@@ -1025,6 +1025,204 @@ func MVCCIterate(engine Engine, startKey, endKey proto.Key, timestamp proto.Time
+// by start and end keys, At each step of the iteration, f() is invoked with the current
+// key/value pair. If f returns true (done) or an error, the iteration stops and the error
+// is propagated.
+func MVCCReverseIterate(engine Engine, startKey, endKey proto.Key, timestamp proto.Timestamp,
```
there is a ton of duplication between this function and MVCCIterate; can you factor out some common code?

spencerkimball replied:
Agreed. This is too big a duplication. Good first step that you wrote it, now I think there's got to be some way to make large parts of the code agnostic about the ordering of the keys and just have two different ways to set up the iteration of key/value pairs.

thundercw replied:
done

petermattis commented on keys/keys.go:
```diff
@@ -355,3 +355,14 @@ func MetaScanBounds(key proto.Key) (proto.Key, proto.Key) {
+		return nil, nil, NewInvalidRangeMetaKeyError("KeyMin or Meta1Prefix can't be used as the key of reverse scan", key)
+	}
+	// Otherwise find the first entry less than or equal to the given key in the same meta prefix.
+	return proto.Key(key[:len(Meta1Prefix)]), key, nil
```
I don't think the cast to `proto.Key` is necessary here (try removing it). Likely this was copy&pasted from elsewhere in this file which contains other unnecessary casts to `proto.Key` (unnecessary because they are redundant, the type being casted is already a `proto.Key`).

thundercw replied:
done

petermattis commented on keys/keys.go:
```diff
@@ -355,3 +355,14 @@ func MetaScanBounds(key proto.Key) (proto.Key, proto.Key) {
+// MetaReverseScanBounds returns the start and end keys of the range within which the
+// desired meta record can be found by means of an engine reverse scan. The given key
+// must be a valid RangeMetaKey as defined by ValidateRangeMetaKey.
+func MetaReverseScanBounds(key proto.Key) (proto.Key, proto.Key, error) {
```
Should add some simple test cases for this function similar to `TestMetaScanBounds`.

thundercw replied:
done

petermattis commented on kv/dist_sender.go:
```diff
@@ -520,6 +526,50 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
 
+// getDescriptorsForReverseScan takes a call and looks up the corresponding range
+// descriptors associated with it. First, the range descriptor for
+// call.Args.EndKey is looked up. If call.Args.Key less than the
```
`s/call.Args.Key less/call.Args.Key is less/g`

thundercw replied:
done

petermattis commented on kv/dist_sender.go:
```diff
@@ -599,9 +662,19 @@ func (ds *DistSender) Send(ctx context.Context, call proto.Call) {
-	}(args.Header().Key)
+	isReverseScan := false
+	revScanArgs, ok := args.(*proto.ReverseScanRequest)
+	if ok {
```
Can simplify this slightly by doing:

```
  revScanArgs, isReverseScan := args.(*proto.ReverseScanRequest)
  if isReverseScan {
    ...
  } else {
    ...
  }
```

thundercw replied:
done

petermattis commented on proto/data.go:
```diff
@@ -75,6 +75,22 @@ func (s KeySlice) Less(i, j int) bool { return s[i].Less(s[j]) }
 func bytesNext(b []byte) []byte {
 	// TODO(spencer): Do we need to enforce KeyMaxLength here?
+	length := len(b)
+	//When the key length exceed KeyMaxLength
```
`s,//When,// When,g`

petermattis commented on proto/data.go:
```diff
@@ -96,6 +112,47 @@ func (k Key) Next() Key {
+	if length == 0 {
+		panic(fmt.Sprint("cannot get the prev bytes of an empty byte array"))
+	}
+	//When the key length exceed KeyMaxLength
```
`s,//When,// When,g`

petermattis commented on proto/internal.proto:
```diff
@@ -43,6 +43,8 @@ message InternalRangeLookupRequest {
   // already in service of pushing intents on meta records. Attempting
   // to resolve intents in this case would lead to infinite recursion.
   optional bool ignore_intents = 3 [(gogoproto.nullable) = false];
+  // Use the reverse scan to pre-fill the range descritor cache.
```
`s/descritor/descriptor/g`

thundercw replied:
done

petermattis commented on storage/engine/engine.go:
```diff
@@ -37,14 +37,21 @@ type Iterator interface {
 	// key range is invalid.
 	Valid() bool
-	// Advances the iterator to the next key/value in the
+	// Next Advances the iterator to the next key/value in the
```
`s/Next Advances/Next advances/g`

thundercw replied:
done

petermattis commented on storage/engine/engine.go:
```diff
@@ -37,14 +37,21 @@ type Iterator interface {
 	// iteration. After this call, the Valid() will be true if the
 	// iterator was not positioned at the last key.
 	Next()
+	// Prev make  the iterator backward to the previous key/value
```
`s/Prev make  the/Prev moves the/g`

thundercw replied:
done

petermattis commented on storage/engine/engine.go:
```diff
@@ -37,14 +37,21 @@ type Iterator interface {
 	// iterator was not positioned at the last key.
 	Next()
+	// Prev make  the iterator backward to the previous key/value
+	// in the iteration. After this call, the Valid() will be true if the
```
`s/the Valid()/Valid()/g`

Should also fix the Next comment which has the same grammatical error.

thundercw replied:
done

petermattis commented on storage/engine/rocksdb/db.h:
```diff
@@ -137,6 +137,11 @@ int  DBIterValid(DBIterator* iter);
 // last key.
 void DBIterNext(DBIterator* iter);
 
+// Make the iterator Backward to the previous key. After this call,
```
`s/Make the iterator Backward/Move the iterator backward/g`

thundercw replied:
done

petermattis commented:
@thundercw Thanks for taking on this work.

@spencerkimball should probably take a look at `MVCCReverseIterate` and the `DistSender` changes.

spencerkimball commented on client/batch.go:
```diff
@@ -308,6 +315,28 @@ func (b *Batch) Scan(s, e interface{}, maxRows int64) {
 	b.initResult(1, 0, nil)
 }
 
+// ReverseScan retrieves the rows between begin (exclusive) and end (inclusive).
```
For clarity, I'd rewrite this as: "in reserve order starting at end (inclusive) and finishing at begin(exclusive)."

spencerkimball commented on kv/dist_sender.go:
```diff
@@ -520,6 +526,50 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
+// descriptors associated with it. First, the range descriptor for
+// call.Args.EndKey is looked up. If call.Args.Key less than the
+// returned descriptor, the next descriptor is obtained as well.
+func (ds *DistSender) getDescriptorsForReverseScan(call proto.Call, options lookupOptions) (*proto.RangeDescriptor, *proto.RangeDescriptor, error) {
```
I think you should remove this method entirely. Too much duplicated code. Just set up an if/else at the top of getDescriptors and create different closures for checking whether another range descriptor is necessary and which key to query:

    needAnotherDescriptorFn := func(desc *proto.RangeDescriptor) bool { return call.Args.Header().Key.Less(desc.StartKey) }
    nextKeyFn := func(desc *proto.RangeDescriptor) proto.Key { return desc.EndKey }

vs.

    needAnotherDescriptorFn := func(desc *proto.RangeDescriptor) bool { return desc.EndKey.Less(call.Args.Header().EndKey) }
    nextKeyFn := func(desc *proto.RangeDescriptor) proto.Key { return desc.StartKey.Prev() }

thundercw replied:
done

spencerkimball commented on proto/api.proto:
```diff
@@ -241,6 +241,25 @@ message ScanResponse {
+  // Must be > 0.
+  optional int64 max_results = 2 [(gogoproto.nullable) = false];
+  optional bool inclusive_key = 3 [(gogoproto.nullable) = false];
+  optional bool inclusive_end_key = 4 [(gogoproto.nullable) = false];
```
Shouldn't inclusive_end_key always be true? I'm not sure it makes sense to provide this option. Also, since we'll end up creating default go structs for the reserve scan request, this will be false by default. That definitely won't give expected results. I think this should just be removed.

Also, seems like maybe we should rename "inclusive_key" to "inclusive_start_key".

thundercw replied:
These two flags are only used internal for cross multiple ranges scan.To user the start key is always exclusive and end key is inclusive.Maybe I need add a `InternalReverseScanRequest` to include the flags and be not exposured to user.

spencerkimball replied:
No, that's not necessary.

spencerkimball commented on storage/range_data_iter.go:
```diff
@@ -136,3 +136,11 @@ func (ri *rangeDataIterator) advance() {
 }
+
+func (r *rangeDataIterator) SeekReverse(key []byte) {
+	panic("TODO(thundercw): unimplemented")
```
I think we'll never want to reverse this iterator. Instead of a TODO, make this:

`panic("cannot reverse scan rangeDataIterator")`

spencerkimball commented on storage/range_data_iter.go:
```diff
@@ -136,3 +136,11 @@ func (ri *rangeDataIterator) advance() {
+}
+
+func (r *rangeDataIterator) Prev() {
+	panic("TODO(thundercw): unimplemented")
```
Ditto here.

thundercw replied:
done

spencerkimball commented:
Nice work! This is a really helpful addition.

I think we need some more unittests though. In particular, we should just add some more unittesting. In particular, a reverse scan across multiple ranges. Look at kv/dist_sender_server_test.go:TestMultiRangeScanDeleteResolve

tschottdorf commented:
the `key.Prev()` situation is a bit unsatisfying. It would be annoying to have to strictly enforce a maximum key length only to have `key.Prev()` work as expected.. is there anything else we can do?

tschottdorf commented on storage/range_command.go:
```diff
@@ -183,6 +185,17 @@ func (r *Range) Scan(batch engine.Engine, args *proto.ScanRequest, reply *proto.
 	return intents
 }
 
+// ReverseScan scans the key range specified by start key through end key up
```
should mention something about the reverse direction.

tschottdorf commented on storage/engine/mvcc.go:
```diff
@@ -1025,6 +1025,204 @@ func MVCCIterate(engine Engine, startKey, endKey proto.Key, timestamp proto.Time
+	return false
+}
+
+// MVCCReverseIterate makes  the iterator backward over the key range specified
```
s/backward/backwards
double space before "the"

thundercw replied:
done

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -527,12 +577,25 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
-		args.Header().EndKey = desc.EndKey
+	if revScanArgs, ok := args.(*proto.ReverseScanRequest); ok {
+		// Truncate the request to our current range, making sure not to
+		// touch it unless we have to (it is illegal to send key on
```
the replacement `EndKey`->`key` botched this comment.

thundercw replied:
updated

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -527,12 +577,25 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
+		if key := args.Header().Key; key != nil && key.Less(desc.StartKey) {
+			defer func(k proto.Key) { args.Header().Key = k }(key)
+			args.Header().Key = desc.StartKey
+			revScanArgs.InclusiveKey = true
```
seems like `InclusiveKey` is tuned towards this multi-range case, but if it's there it should probably also work for scans on a single range. Currently it doesn't because you overwrite it below. Maybe it would be better if the semantics for `ReverseScan` would be that it excludes the `EndKey`, but always includes `StartKey`? Then you have less trouble, and clients that want the opposite behaviour can simply scan from `EndKey.Next()` to `StartKey.Next()`.

thundercw replied:
That's a good idea. It indeed can simplify the work. I will update the logic according to your suggestions.Thanks.

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -520,6 +526,50 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
+	if revScanRequest.InclusiveEndKey {
+		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().EndKey, options)
+	} else {
+		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().EndKey.Prev(), options)
```
I'm very interested in getting rid of the `Prev()` here. The `rangeCache` is a tree, it has both `Ceil()` and `Floor()` so the only problem is that `Floor()` gives you the wrong thing for a key which is in the tree (because it stops on equality). But you can wrap the key in a way that makes it never compare equal to anything else, and then flooring on that should give you the desired previous descriptor (if cached, but you can check easily if what you get back matches).
That's a bit of complication but can probably be wrapped up nicely. I just feel that reintroducing `Prev()` causes a lot of tricky issues.

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -520,6 +526,50 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
+		}
+		// This next lookup is likely for free since we've read the
+		// previous descriptor and range lookups use cache prefetching.
+		descNext, err = ds.rangeCache.LookupRangeDescriptor(desc.StartKey.Prev(), options)
```
ditto.

thundercw replied:
done

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=1686) <br/>You should sign our Contributor License Agreement in order to get your pull request merged.<br/>

thundercw commented:
@spencerkimball ,@petermattis, @tschottdorf  , @tamird    
The major changes are as follows:
1.The semantics for ReverseScan is changed from `(startkey,endky]` to `[startkey,endkey) ` .
2. The key.Prev() has been removed again. Add a reverse flag in `LookupRangeDescriptor  lookupOptions` to support look up range for reverse scan  .
3. reconstructed the MVCCIterate func to support scan and reverse scan .
4. Add some comments and unit tests.

Please take another look!  thanks.

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -526,11 +548,19 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
-		args.Header().EndKey = desc.EndKey
+	// touch it unless we have to
+	if _, ok := args.(*proto.ReverseScanRequest); ok {
+		// If the revsese scan request is cross multiple ranges.
```
s/revsese/reverse/

tschottdorf commented on kv/range_cache.go:
```diff
@@ -202,22 +203,29 @@ func (rdc *rangeDescriptorCache) EvictCachedRangeDescriptor(descKey proto.Key, s
+	// end-key is non-inclusive.
+	var metaKey proto.Key
+	if !isReverse {
+		// If inclusive is false, we access the cache using key.Next().
```
comment needs update.

thundercw replied:
updated

tschottdorf commented on proto/api.proto:
```diff
@@ -241,6 +241,21 @@ message ScanResponse {
+// start(inclusive) and end(exclusive) keys for the scan and the maximum number of results.
+message ReverseScanRequest {
+  optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  // Must be > 0.
```
why can't this be zero for an unbounded number of results?

thundercw replied:
Updated.

tschottdorf commented on storage/engine/mvcc.go:
```diff
@@ -924,19 +924,97 @@ func MVCCDeleteRange(engine Engine, ms *MVCCStats, key, endKey proto.Key, max in
+func MVCCReverseScan(engine Engine, key, endKey proto.Key, max int64, timestamp proto.Timestamp,
+	consistent bool, txn *proto.Transaction) ([]proto.KeyValue, []proto.Intent, error) {
+	res := []proto.KeyValue{}
+	intents, err := MVCCIterate(engine, key, endKey, timestamp, consistent, txn, true,
```
if the only difference between `MVCC{,Reverse}Scan` is the `true` here, consider unifying them more.

thundercw replied:
done

tschottdorf commented on storage/engine/mvcc.go:
```diff
@@ -924,19 +924,97 @@ func MVCCDeleteRange(engine Engine, ms *MVCCStats, key, endKey proto.Key, max in
-		}
-		return false, nil
-	})
+	intents, err := MVCCIterate(engine, key, endKey, timestamp, consistent, txn, false,
```
try replacing `false` by `false /* reverse */`. That way the reader knows what it does.

thundercw replied:
done

tschottdorf commented on storage/engine/rocksdb/db.h:
```diff
@@ -137,6 +137,11 @@ int  DBIterValid(DBIterator* iter);
 // last key.
 void DBIterNext(DBIterator* iter);
 
+// Moves the iterator Backward to the previous key. After this call,
```
s/Backward/back/

thundercw replied:
done

tschottdorf commented on storage/range_raftstorage.go:
```diff
@@ -92,7 +92,8 @@ func (r *Range) Entries(lo, hi, maxBytes uint64) ([]raftpb.Entry, error) {
 		keys.RaftLogKey(r.Desc().RaftID, hi),
-		proto.ZeroTimestamp, true /* consistent */, nil /* txn */, scanFunc)
+		proto.ZeroTimestamp,
+		true /* consistent */, nil /* txn */, false /*reberse */, scanFunc)
```
s/reberse/reverse/

thundercw replied:
done

tschottdorf commented on storage/store.go:
```diff
@@ -478,34 +478,35 @@ func (s *Store) Start(stopper *stop.Stopper) error {
-		_, suffix, _ := keys.DecodeRangeKey(kv.Key)
-		if !suffix.Equal(keys.LocalRangeDescriptorSuffix) {
+	if _, err := engine.MVCCIterate(s.engine, start, end, now, false /* !consistent */, nil,
+		false /*reverse*/, func(kv proto.KeyValue) (bool, error) {
```
`/* !reverse */`

thundercw replied:
done

tschottdorf commented:
@thundercw sorry for letting this sit for so long. I've browsed the change (without going into technical correctness) and it looks fairly clean. I'm especially happy that you managed to get rid of `key.Prev()` and that the logic in most places hasn't complicated terribly.
Would you mind rebasing this? I'll give it a close look after.
Thanks!

thundercw commented:
Sure

thundercw commented:
@tschottdorf  Rebased , PTAL!

tamird commented on client/batch.go:
```diff
@@ -308,6 +315,29 @@ func (b *Batch) Scan(s, e interface{}, maxRows int64) {
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (b *Batch) ReverseScan(s, e interface{}, maxRows int64) {
```
seems like this can be done with less duplication of code from `Scan`.

thundercw replied:
Now the semantics of ReverseScan is same with Scan(start key inclusive and end key exclusive) except seeking order.  Is it better to add a reverse flag in proto.ScanRequest instead of adding a new ReverseScanRequest ?

thundercw replied:
Adding a new ReverseScanRequest will be easy to use because it is clearer than a flag. But it will lead to some duplicated code .

tschottdorf replied:
Would we lose anything if we allowed `Scan` requests with `EndKey < Key` and treated those as reverse scans? Seems economical and easy to understand: The first key is inclusive, and then you scan towards the second. Unless that has subtle issues, I think I would like that.

tschottdorf replied:
It would definitely have some not-so-subtle issues, namely that it requires an update to our routing. Probably not a good idea internally, but for client usage could still make sense - but that's less important now.

An unrelated thought though: Some of the complexity seems to stem from the fact that with a `ReverseScan` on `"a"`-`"z"`, the `EndKey` `"z"` is inclusive. If we implemented only `EndKey`-exclusive `ReverseScan` internally, that would be easier (because it plays along with `EndKey` being excluded in range operations) and we can easily expose the inclusive one by calling `Next()` on the EndKey for clients. Have you considered that?

thundercw replied:
Yes ,I have considered that. In some SQL  engine like sqlite4, the end key of the reverse scan is always inclusive, and expose key.Next() to it is not a good idea. I think it is better to change the semantics of reverse scan to start key exclusive and end key inclusive to client interface. But internally our implementation is opposite by conversion the key and end key before send in kv/db.go. What`s your opinion?

tschottdorf replied:
> I think it is better to change the semantics of reverse scan to start key exclusive and end key inclusive to client interface

I'm getting confused with all the keys here. Do you mean `("a","b")` (left and right side excluded)? That's what I was suggesting. Scanning `("a", "b"]` (left-exclude, right-include) is just calling `Next()` on the right key.

thundercw replied:
For user/client , Scanning `("a", "b"]` (left-exclude, right-include) 
In our internally implementation , Scanning `["a".next(), "b".next())`  (left-include, right-exclude);
The conversion place is before `Send()` in kv/db.go .

tschottdorf replied:
I think I've confused myself right now. Let me summarize.
whether the left endpoint is included or not doesn't matter. It's all about `EndKey`. I'm assuming included below.
Currently:
* `[a-z)` (right-exclude)
* descriptor lookup: needs `.Next()` because the reverse scan is exclusive

the alternative, including the right endpoint, is awkward because that's not how our key ranges work.

So I think you've indeed implemented the one that makes most sense.

tschottdorf replied:
And if someone wants right-include, they can always call with `EndKey.Next()`. For left-exclude, `StartKey.Next()`. So that implemention includes all others.

tschottdorf replied:
just got your message above. Sounds good. I wouldn't bother with the client implementation now - just go with what you have (left-incl, right-excl for everybody), it's easy to change later, for now better if all our reverse scan related functions do the same.

thundercw replied:
OK. The main logic will not be changed.

tamird commented on client/client_test.go:
```diff
@@ -428,6 +428,35 @@ func TestClientBatch(t *testing.T) {
+
+	revScan1 := b.Results[0].Rows
+	revScan2 := b.Results[1].Rows
+	if len(revScan1) != 5 || len(revScan2) != 5 {
```
can you extract this `5` into a local?

thundercw replied:
done

tamird commented on client/client_test.go:
```diff
@@ -428,6 +428,35 @@ func TestClientBatch(t *testing.T) {
+			len(revScan1), len(revScan2))
+	}
+	for i := 0; i < 5; i++ {
+		if key := proto.Key(revScan1[i].Key); !key.Equal(keys[4-i]) {
```
ditto with this `4` and `9` below

thundercw replied:
done

tamird commented on client/db.go:
```diff
@@ -318,6 +318,20 @@ func (db *DB) Scan(begin, end interface{}, maxRows int64) ([]KeyValue, error) {
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (db *DB) ReverseScan(begin, end interface{}, maxRows int64) ([]KeyValue, error) {
```
this also can share code with Scan

thundercw replied:
done /g

tamird commented on client/txn.go:
```diff
@@ -208,6 +208,20 @@ func (txn *Txn) Scan(begin, end interface{}, maxRows int64) ([]KeyValue, error)
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (txn *Txn) ReverseScan(begin, end interface{}, maxRows int64) ([]KeyValue, error) {
```
and here

tamird commented on keys/keys.go:
```diff
@@ -357,3 +357,17 @@ func MetaScanBounds(key proto.Key) (proto.Key, proto.Key) {
+// must be a valid RangeMetaKey as defined by ValidateRangeMetaKey.
+func MetaReverseScanBounds(key proto.Key) (proto.Key, proto.Key, error) {
+	if key.Equal(proto.KeyMin) || key.Equal(Meta1Prefix) {
+		return nil, nil, NewInvalidRangeMetaKeyError("KeyMin or Meta1Prefix can't be used as the key of reverse scan", key)
```
s/or/and/

thundercw replied:
done

tamird commented on keys/keys.go:
```diff
@@ -357,3 +357,17 @@ func MetaScanBounds(key proto.Key) (proto.Key, proto.Key) {
+	if key.Equal(proto.KeyMin) || key.Equal(Meta1Prefix) {
+		return nil, nil, NewInvalidRangeMetaKeyError("KeyMin or Meta1Prefix can't be used as the key of reverse scan", key)
+	}
+	if key.Equal(Meta2Prefix) {
```
can you add a comment here to mirror the comment in `MetaScanBounds`?

thundercw replied:
done

tamird commented on kv/dist_sender.go:
```diff
@@ -531,11 +553,19 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
-	if endKey := args.Header().EndKey; endKey != nil && !endKey.Less(desc.EndKey) {
-		defer func(k proto.Key) { args.Header().EndKey = k }(endKey)
-		args.Header().EndKey = desc.EndKey
+	// touch it unless we have to
```
missing a period

tamird replied:
also why did you remove the rest of this comment?

tamird replied:
ah, I see it moved.

tamird commented on kv/dist_sender.go:
```diff
@@ -531,11 +553,19 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
-		defer func(k proto.Key) { args.Header().EndKey = k }(endKey)
-		args.Header().EndKey = desc.EndKey
+	// touch it unless we have to
+	if _, ok := args.(*proto.ReverseScanRequest); ok {
```
nit: name this `isReverseScan`?

thundercw replied:
done

tamird commented on kv/range_cache.go:
```diff
@@ -228,7 +236,14 @@ func (rdc *rangeDescriptorCache) getCachedRangeDescriptorLocked(key proto.Key) (
 	// Check that key actually belongs to range
 	if !rd.ContainsKey(keys.KeyAddress(key)) {
-		return nil, nil
+		if !isReverse {
```
i don't understand this change; does the reverse scan change the edge key inclusion rule?

thundercw replied:
No, the edge key inclusion rule is not changed. Because reverse scan is seek backwards from endkey(exclusive) to startkey and if the end key is the range's start key, we need to get the previous range desc.

tamird commented:
LGTM though @petermattis should look at the engine changes

tschottdorf commented on keys/keys.go:
```diff
@@ -357,3 +357,17 @@ func MetaScanBounds(key proto.Key) (proto.Key, proto.Key) {
+		return Meta1Prefix, Meta1Prefix.PrefixEnd(), nil
+	}
+	// Otherwise find the first entry greater than the given key in the same meta prefix.
+	return key[:len(Meta1Prefix)], key.Next(), nil
```
Might be helpful to give an example here:

If we have ranges `["", "f")` and `["f", "z")`, then we'll have corresponding meta records at `"f"` and `"z"`. If you're looking for the meta record for key `"f"`, then you want the second record (keyed under `"z"`), hence `key.Next()` below.

tschottdorf replied:
With my suggestion of implementing `EndKey`-exclusive `ReverseScan`, the `Next()` here would be dropped.

thundercw replied:
The rocksdb.Seek() will position the iterator at the first key that is >= "key". If we have ranges ["", "f") , ["f", "z") and your reverse scan range is `["a", "f")`, you will use `"f"` to look up the range. `"f"` is the key of the first meta row and the iterator will position it if we does not use `"f".Next()` .So the target range will be ignored by `Endkey` -exclusive .

tschottdorf replied:
I don't understand. In your scenario, the range I would want to hit with end exclusive reverse scan is `["", "f")`. Its descriptor is at `"f"` and that's where the iterator starts if we drop the `Next()`. Am I missing the point?

thundercw replied:
In this scenario, if we drop the next(), "f" is set to the endkey of the `MvccReverseScan`  and the end key is exclusive.

tschottdorf replied:
ah, I see. Thanks for the explanation.

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -492,7 +494,27 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
+		needAnotherDescriptorFn = func(desc *proto.RangeDescriptor) bool {
+			return desc.EndKey.Less(call.Args.Header().EndKey)
+		}
+		nextKeyFn = func(desc *proto.RangeDescriptor) proto.Key { return desc.EndKey }
```
could you fuse `needAnotherDescriptorFn` and `nextKeyFn` into
```go
needAnother = func(desc *proto.RangeDescriptor) (bool, proto.Key) { }?
```

thundercw replied:
done

tschottdorf commented on kv/db_test.go:
```diff
@@ -114,6 +114,19 @@ func TestKVDBCoverage(t *testing.T) {
 		}
 	}
 
+	// Test reverse scan
```
punctuation.

thundercw replied:
done

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -531,11 +553,19 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
-		args.Header().EndKey = desc.EndKey
+	// touch it unless we have to
+	if _, ok := args.(*proto.ReverseScanRequest); ok {
+		// If the reverse scan request is cross multiple ranges.
```
> // Truncate to the current Range if the reverse scan crosses range boundaries.

thundercw replied:
done /g

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -531,11 +553,19 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
+		if endKey := args.Header().EndKey; endKey != nil && !endKey.Less(desc.EndKey) {
+			defer func(k proto.Key) { args.Header().EndKey = k }(endKey)
+			args.Header().EndKey = desc.EndKey
+		}
```
why do we truncate the request here at all again? This is `sendAttempt`, I thought it already got a pre-truncated args. See whether this is really necessary (possibly the old code snuck in during a refactor).

tschottdorf replied:
ok, I think the code is here because doing it outside of `sendAttempt` (where it belongs semantically) can't `defer()` as elegantly. Can you still pull this up? I'm commenting at the call site below.

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -635,6 +671,12 @@ func (ds *DistSender) Send(ctx context.Context, call proto.Call) {
 			}
 			// At this point reply.Header().Error may be non-nil!
 			curReply, err = ds.sendAttempt(trace, args, desc)
```
as per my comment above, would do the truncation here:

```go
func() {
    defer func() {
        // untruncate
    }()
    // truncate
    curReply, err = ds.sendAttempt(/* ... */)
}()
```

thundercw replied:
Can I refactor the code by creating a new issue after my current PR is merged ?

tschottdorf replied:
yes.

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -600,10 +630,16 @@ func (ds *DistSender) Send(ctx context.Context, call proto.Call) {
-	defer func(key proto.Key) {
-		args.Header().Key = key
-	}(args.Header().Key)
+	_, isReverseScan := call.Args.(*proto.ReverseScanRequest)
```
Add a comment similar to line 624f here.

tschottdorf commented on storage/replica_command.go:
```diff
@@ -398,13 +414,29 @@ func (r *Replica) InternalRangeLookup(batch engine.Engine, args proto.InternalRa
+		kvs, intents, err = engine.MVCCScan(batch, startKey, endKey, rangeCount,
+			args.Timestamp, consistent, args.Txn)
+	} else {
+		// We want to search for the metadata key just less than args.Key. Reverse
```
currently, it's not "just less": you're searching "less or equal" (but see my suggestion above).

thundercw replied:
done

tschottdorf commented on storage/store.go:
```diff
@@ -478,34 +478,35 @@ func (s *Store) Start(stopper *stop.Stopper) error {
-		// Only consider range metadata entries; ignore others.
-		_, suffix, _ := keys.DecodeRangeKey(kv.Key)
-		if !suffix.Equal(keys.LocalRangeDescriptorSuffix) {
+	if _, err := engine.MVCCIterate(s.engine, start, end, now, false /* !consistent */, nil,
```
maybe annotate the `nil` with a `/* */` comment as well while you're there.

thundercw replied:
done

tschottdorf commented on storage/store.go:
```diff
@@ -478,34 +478,35 @@ func (s *Store) Start(stopper *stop.Stopper) error {
+			// TODO(bdarnell): Also initialize raft groups when read leases are needed.
+			// TODO(bdarnell): Scan all ranges at startup for unapplied log entries
+			// and initialize those groups.
 			return false, nil
```
nothing changed in the above block except you adding the `false`, correct?

thundercw replied:
Yes

tschottdorf commented:
added another round of comments. Could you add any further updates as separate commits (we can squash later)? That'll make it easier to keep track of changes.

thundercw commented:
OK

tschottdorf commented:
@thundercw can you acknowledge all of the comments you haven't responded to yet (also old ones - just add a "done" comment when you've addressed them in a commit you're going to push)? Once that's through, I'll give it a last scroll through and then'll merge. Thanks for the good work!

thundercw commented:
Sure.

thundercw commented:
@tschottdorf  All the comments have been replied. New commits have been revised according to the comments. Additionally, I fixed two bugs and added some tests.
PTAL! Thanks.

tschottdorf commented:
@thundercw good catch. Some more comments - really with a change as big as that we're never going to run out of them. I say you fix the easy ones (comment suggestions, spelling, ...) and then squash and we merge that. You can address tests, refactor, ... in a small follow-up PR, as I will likely do too. Ping me when you're ready.

tschottdorf commented:
and, again, thanks. This is a pretty involved contribution.

thundercw commented:
@tschottdorf  I have updated per your comments. Please have a look and merge it.
Thanks for your time on this PR.

tschottdorf commented:
:tada:

spencerkimball commented:
Nice work, @thundercw!
